bpo-35936: Updates to modulefinder

brandtbucher opened the pull request:
This patch includes a few useful fixes to `Lib/modulefinder.py`. Namely:

- It doesn't crash if it encounters a syntax error ([bpo-17396](https://bugs.python.org/issue17396)).
- It doesn't report certain name collisions as bad ([bpo-35376](https://bugs.python.org/issue35376)).
- It doesn't use mutable default arguments in its initializer.
- Most importantly, it now uses `importlib` instead of `imp`, which is deprecated ([bpo-25160](https://bugs.python.org/issue25160), [bpo-20020](https://bugs.python.org/issue20020)) As a benefit, frozen built-in modules (`zipimport`, for example) are now correctly reported.

With the exception of these bug fixes, I've been very careful to preserve the original behavior, and have not changed any part of the API. This patch also includes 2 new regression tests.

<!-- issue-number: [bpo-35936](https://bugs.python.org/issue35936) -->
https://bugs.python.org/issue35936
<!-- /issue-number -->

brandtbucher committed:
Replace deprecated imp usage.

Constants imported from imp have been moved to private module-level constants, and ModuleFinder.find_module has been refactored to use importlib. Other than an improvement on how frozen builtin imports are reported (as the frozen imports they are, rather than the stdlib modules they *may* have originated from), these changes maintain complete compatibility with past versions... including odd behavior for returning relative (below current directory, but not a C extension) vs. absolute (above current directory, or a C extension) paths.

brandtbucher committed:
Properly handle SyntaxErrors in Python source files.

SyntaxErrors in the target module will rise normally, while SyntaxErrors in dependencies will be added to badmodules. This includes a new regression test.

brandtbucher committed:
Replace mutable default values.

Bound empty lists have been replaced with the "if param is None" idiom.

brandtbucher committed:
Fix name collision bug.

This fixes an issue where a "fromlist" import with the same name as a previously failed import would be incorrectly added to badmodules. This includes a new regression test.

brandtbucher committed:
Fixed whitespace.

Whoops.

brandtbucher committed:
Handle differing drives on Windows.

Specifically, when checking whether the returned path should be relative or absolute.

brandtbucher committed:
Again, fixed whitespace.

Whoops.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher committed:
Fix relative imports.

Hopefully, this will help the failing tests on Windows.

pganssle commented on Misc/NEWS.d/next/Library/2019-02-08-04-46-58.bpo-35936.q2ArjU.rst:
```diff
@@ -0,0 +1,2 @@
+:mod:`modulefinder` has been refactored to use the non-deprecated :mod:`importlib`. Additionally, several long-standing bugs have been fixed.
+Patch by Brandt Bucher.
```
I personally do not like it when a changelog makes references to bugfixes without specifying what they are. I think maybe it would be appropriate to add separate news entries for each of the long-standing bugs that were fixed. I don't think there's any rule that says that a given PR must have only one changelog entry, and presumably it would be nice to have one changelog ticket per BPO issue number.

brandtbucher replied:
I think that makes sense, especially if the `imp` fix is eventually backed out.

brandtbucher committed:
Fix import side-effects.

importlib.util.find_spec(fullname, parent), where fullname contains a dot, results in the parent module being imported. This fix ensures that children are correctly located while also avoiding import side-effects by temporarily clearing sys.modules and modifying sys.path. Now, neither a dotted name nor parent are required.

brandtbucher committed:
Wait for sys.path to update.

I have a hunch the failing tests are due to a race condition in how sys.path updates. Sleep a bit here to see.

brandtbucher committed:
Bump sleep time.

More test are passing, which supports my theory.

brandtbucher committed:
WIP: Replace sleep with busy wait.

*crosses fingers*

brandtbucher committed:
Remove busy wait.

I guess that wasn't the issue...

brandtbucher committed:
Remove NEWS entry.

This will be replaced with several more specific entries detailing the various bug fixes.

brandtbucher committed:
Mutate sys.path in-place.

Perhaps there is some underlying reference issue with straight-up sys.path replacement. Once again, fingers crossed!

brandtbucher committed:
Enter RLock during import search.

Perhaps test threads are stomping on each others' sys.path?

brandtbucher committed:
Try multiprocessing.RLock instead of threading.

These are processes, after all...

brandtbucher committed:
Try a couple of possible fixes for failing lookups.

This is a longshot, but it's pretty much all I have left in terms of ideas.

brandtbucher committed:
Remove sys._clear_type_cache calls.

Is it a refleak issue?

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher committed:
Break out new code into standalone function.

I've also added helpful documentation outlining the reasoning behind various design decisions.

brandtbucher committed:
Don't modify import state in-place.

Changing references seems like a safer option and doesn't require extra copy operations.

brandtbucher committed:
Reorder new import.

Minor. Keep our imports in ASCIIbetical order.

brandtbucher committed:
Minor style stuff.

Remove a comment, change a docstring, and add an ACK.

ncoghlan commented on Lib/modulefinder.py:
```diff
@@ -43,6 +52,87 @@ def ReplacePackage(oldname, newname):
+    """An importlib reimplementation of imp.find_module (for our purposes)."""
+
+    # To correctly find the module on the given path,  we need to
+    # temporarily clear the import state of the program and modify sys.path.
```
While this is a true statement, it's also substantial scope creep for `modulefinder`, which has historically only worked for regular filesystem paths. So I think it would be good to break this aspect of the change up into two parts:

1. In this PR, use [importlib.machiner.PathFinder.find_spec](https://docs.python.org/3/library/importlib.html#importlib.machinery.PathFinder) *without* clearing any of the import system caches to emulate the `imp.find_module` behaviour

2. File a follow-up enhancement request to expand module finder to respect `sys.meta_path` hooks

ncoghlan replied:
I realised I left out why the change concerns me and I'd like to see it reviewed separately: the actual `imp.find_module` implementation doesn't have any side effects, so calling it from multiple threads is safe. Rebinding `sys` module attributes on the other hand is decidedly *un*safe, so if that enhancement is going to happen, we'll need to find a different way of doing it that doesn't have global side effects :)

brandtbucher committed:
Replace importlib.util.find_spec with importlib.machinery.PathFinder.find_spec.

This swap avoids behavior changes and concurrency issues. The reverted changes will be proposed and reviewed separately. Thanks @ncoghlan!

brandtbucher committed:
Invalidate caches before find_spec call.

It appears that we're still hitting some caches. This *should* help...

brandtbucher committed:
Fix whitespace.

Whoops.

brandtbucher committed:
Remove duplicate-bpo NEWS entries.

These will be replaced with new ones properly referencing their respective bpo issue numbers.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher committed:
Empty commit.

This is to force a rebuild on the PR.

ncoghlan commented on Lib/modulefinder.py:
```diff
@@ -69,16 +138,16 @@ def __repr__(self):
 class ModuleFinder:
 
-    def __init__(self, path=None, debug=0, excludes=[], replace_paths=[]):
+    def __init__(self, path=None, debug=0, excludes=None, replace_paths=None):
```
While this seems like a good change to me, it also seems worth mentioning in one of the NEWS entries (probably by adding it to https://github.com/python/cpython/pull/11787/files#diff-0c343f3267eee48d7cb592115947061b ).

ncoghlan reviewed (commented):
One minor release docs comment, but otherwise this version looks good to me. However, I'd really like @ronaldoussoren to take another look before we merge it.

brandtbucher committed:
Remove NEWS entry.

This is going to have info about the mutable defaults added.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+
+    importlib.machinery.PathFinder.invalidate_caches()
+
+    spec = importlib.machinery.PathFinder.find_spec(name, path+sys.path)
```
``(path+sys.path)`` changes the semantics that existed before this patch (_find_module might find modules on ``sys.path`` that aren't found on ``path``).

Why is it necessary to include ``sys.path`` on the search path here?

brandtbucher replied:
Appending `sys.path` is left over from my early work on this change.. and I don't actually remember what led me to add it there. Either way, I agree that it is no longer necessary.

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+def _find_module(name, path=None):
+    """An importlib reimplementation of imp.find_module (for our purposes)."""
+
+    importlib.machinery.PathFinder.invalidate_caches()
```
I'd add a comment here that explains why clearing the cache is necessary, the need is not immediately obvious.

brandtbucher replied:
Thanks. I've clarified this bit in the source.

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+    if spec is None:
+        raise ImportError("No module named {name!r}".format(name=name), name=name)
+
+    if isinstance(spec.loader, type):  # Some special cases:
```
Not sure if using ``issubclass`` necessary, just testing for these two special cases is good enough:

```
if spec.loader is import lib.machinery.BuiltinImporter:
  ...
```

Another special case is the loader for implicit namespace packages (which is currently an instance of a private class, see bpo-35673). Those are probably not handled at all at the moment by this module.

brandtbucher replied:
I've refactored these to use identity checking, as you suggested. Much cleaner.

I think it makes the most sense to wait on bpo-35673's resolution before trying to hack our way around namespace packages here... that could be a separate PR. What are your thoughts @ronaldoussoren?

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+
+        # If the path is a descendant of the CWD, it is reported as a relative filepath
+        # ...unless it is an extension, in which case it will be absolute (see below)
+        # ...otherwise, it is fully resolved.
```
I wonder if this behaviour should be emulated, although there's bound to be someone that relies on the current behaviour.

brandtbucher replied:
Good news: it looks like `PathFinder` does this weird path formatting for us! So either way, no need for this ugly code.

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+        pass
+
+    base, suffix = os.path.splitext(file_path)
+
```
I'm not sure about the code that follows. The *kind* can be deduced from the type of loader (at least for loaders in the stdlib).

brandtbucher replied:
I've replaced extension checking with `isinstance` checks against the concrete loaders in `importlib.machinery`. I agree that it appears clearer and more robust than `imp`-style suffix checks.

ronaldoussoren commented on Lib/modulefinder.py:
```diff
@@ -43,6 +51,67 @@ def ReplacePackage(oldname, newname):
+    elif suffix in importlib.machinery.SOURCE_SUFFIXES:
+
+        if name != "__init__" and os.path.basename(base) == "__init__":
+            return (None, os.path.dirname(file_path), ("", "", _PKG_DIRECTORY))
```
- Loader has a method ``is_package()`` that does this test. 

- The package ``__init__`` can also be an extension (although I'm not aware of any code that actually does this)

brandtbucher replied:
Thanks, didn't know about `is_package`. That makes it easy to move this check out in front of this branch.

brandtbucher committed:
Remove loader subclass checks.

Identity checking is sufficient here.

brandtbucher committed:
Use spec.loader for package checking.

This is more robust than using a string comparison with "__init__".

brandtbucher committed:
Remove sys.path from path.

This isn't actually needed (or wanted) here.

brandtbucher committed:
Use loader type to classify imports.

This is probably a better idea than using the extension alone.

brandtbucher committed:
Remove absolute/relative path formatting.

Interestingly, it looks like PathFinder does this for us! That's great, because it really cleans up this section a lot.

brandtbucher committed:
Reorder branching.

Hopefully we can get a tiny boost here by checking likely outcomes first.

brandtbucher committed:
Add comments on cache invalidation.

This makes the unintuitive reasoning behind this call explicit.

ncoghlan reviewed (approved):
Same situation as before - while I'm personally happy with the way the importlib APIs are now being used, @ronaldoussoren is the main reviewer from the actual `modulefinder` functionality side.
